File: linux-client/checkupdate.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// CheckUpdateResponse 检查更新响应
type CheckUpdateResponse struct {
	HasUpdate     bool   `json:"has_update"`
	CurrentVersion string `json:"current_version"`
	NewVersion    string `json:"new_version"`
	DownloadURL   string `json:"download_url"`
	Hash          string `json:"hash"`
	Size          int64  `json:"size"`
	Filename      string `json:"filename"`
}

// CheckForUpdate 检查是否有新版本
func CheckForUpdate() (*CheckUpdateResponse, error) {
	token := tokenManager.GetShortToken()
	if token == "" {
		return nil, fmt.Errorf("no token available")
	}

	// 构建请求 URL
	url := fmt.Sprintf("%s/api/client/check-update", config.ServerURL)

	// 创建 HTTP 请求
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("create request failed: %v", err)
	}

	// 设置 Token
	req.Header.Set("X-Client-Token", token)

	// 【关键】携带本地实际运行版本，供后端比对并同步更新记录，避免更新后版本不同步导致重复更新
	req.Header.Set("X-Client-Version", version)

	// 发送请求
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("HTTP request failed: %v", err)
	}
	defer resp.Body.Close()

	// 读取响应
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response failed: %v", err)
	}

	// 检查状态码
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("check update failed: HTTP %d, body: %s", resp.StatusCode, string(body))
	}

	// 解析响应
	var result CheckUpdateResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse response failed: %v", err)
	}

	log.Printf("✅ 版本检查结果：当前=%s, 最新=%s, has_update=%v", 
		result.CurrentVersion, result.NewVersion, result.HasUpdate)

	return &result, nil
}

// checkUpdateLoop 版本检查循环
func checkUpdateLoop() {
	log.Println("Starting version check loop (every 5 minutes)...")

	// 立即执行第一次检查
	checkAndDownloadUpdate()

	// 之后每 5 分钟检查一次
	ticker := time.NewTicker(5 * time.Minute)
	defer ticker.Stop()

	for range ticker.C {
		checkAndDownloadUpdate()
	}
}

// checkAndDownloadUpdate 检查并下载更新
func checkAndDownloadUpdate() {
	log.Println("[UPDATE] Checking for new version...")

	response, err := CheckForUpdate()
	if err != nil {
		log.Printf("[UPDATE] Check failed: %v", err)
		return
	}

	if !response.HasUpdate {
		log.Println("[UPDATE] No new version available")
		return
	}

	log.Printf("[UPDATE] New version found: %s -> %s", response.CurrentVersion, response.NewVersion)

	// 【关键】本地防护：目标版本必须高于本地实际版本才安装，
	// 避免服务端最新包版本低于本地时被降级安装旧包
	if !isNewerVersion(response.NewVersion, version) {
		log.Printf("[UPDATE] Target version %s is not newer than local version %s, skip installation", response.NewVersion, version)
		return
	}
	
	// 调用下载器
	tempPath, err := DownloadUpdate(response.DownloadURL, response.Filename, response.Hash)
	if err != nil {
		log.Printf("[UPDATE] Download failed: %v", err)
		return
	}

	log.Println("[UPDATE] Update package downloaded successfully")

	// 调用安装器进行安装
	if err := InstallUpdate(tempPath); err != nil {
		log.Printf("[UPDATE] Installation failed: %v", err)
		return
	}

	log.Println("[UPDATE] ✅ Update installed successfully!")
}

// isNewerVersion 语义化版本比较：v1 > v2 返回 true（点分数字段逐段比较）
func isNewerVersion(v1, v2 string) bool {
	p1 := strings.Split(v1, ".")
	p2 := strings.Split(v2, ".")
	maxLen := len(p1)
	if len(p2) > maxLen {
		maxLen = len(p2)
	}
	for i := 0; i < maxLen; i++ {
		var n1, n2 int
		if i < len(p1) {
			n1, _ = strconv.Atoi(p1[i])
		}
		if i < len(p2) {
			n2, _ = strconv.Atoi(p2[i])
		}
		if n1 != n2 {
			return n1 > n2
		}
	}
	return false // 版本相等
}

File: windows-client/src/checkupdate.rs
//! Windows 客户端版本检查模块

use crate::{api, config::Config, token::TokenManager, downloader, installer};
use std::thread;
use std::time::Duration;

/// 检查更新响应结构 (与服务端 API 匹配)
#[derive(Debug, serde::Deserialize)]
pub struct CheckUpdateResponse {
    pub has_update: bool,
    pub current_version: String,
    pub new_version: String,
    pub download_url: String,
    pub hash: String,
    #[serde(default)]
    pub size: i64,
    pub filename: String,
}

/// 版本检查间隔：5 分钟
const UPDATE_CHECK_INTERVAL: Duration = Duration::from_secs(5 * 60);

/// 检查是否有新版本
pub fn check_for_update(config: &Config, token_manager: &TokenManager) -> Result<CheckUpdateResponse, String> {
    log::info!("[UPDATE] Checking for new version...");
    
    let short_token = token_manager.short_token();
    if short_token.is_empty() {
        return Err("Failed to get short token".to_string());
    }
    
    api::check_update_blocking(&config.server_url, short_token)
}

/// 语义化版本比较：v1 > v2 返回 true（点分数字段逐段比较）
fn is_newer_version(v1: &str, v2: &str) -> bool {
    let p1: Vec<u64> = v1.split('.').filter_map(|s| s.parse().ok()).collect();
    let p2: Vec<u64> = v2.split('.').filter_map(|s| s.parse().ok()).collect();
    let max_len = p1.len().max(p2.len());
    for i in 0..max_len {
        let n1 = p1.get(i).copied().unwrap_or(0);
        let n2 = p2.get(i).copied().unwrap_or(0);
        if n1 != n2 {
            return n1 > n2;
        }
    }
    false // 版本相等
}

/// 版本检查循环（非阻塞 - 立即启动后台线程）
pub fn version_check_loop(
    config: Config,
    token_manager: TokenManager,
) -> Result<(), String> {
    log::info!("[UPDATE] Starting automatic update checker...");
    
    // 启动后台定时线程（首次立即检查 + 每 5 分钟定时检查）
    let config_clone = config.clone();
    let token_manager_clone = token_manager.clone();
    
    thread::spawn(move || {
        // 1. 立即执行第一次检查
        log::info!("[UPDATE] Performing initial version check...");
        match run_check_and_install(&config_clone, &token_manager_clone) {
            Ok(true) => log::info!("[UPDATE] Initial update installed successfully"),
            Ok(false) => {} // 没有更新或失败，继续循环
            Err(e) => log::warn!("[UPDATE] Initial check failed: {}", e),
        }
        
        let interval = UPDATE_CHECK_INTERVAL;
        
        loop {
            // 等待下一个检查时间
            std::thread::sleep(interval);
            
            log::info!("[UPDATE] Performing scheduled version check...");
            
            match run_check_and_install(&config_clone, &token_manager_clone) {
                Ok(true) => log::info!("[UPDATE] Update installed successfully"),
                Ok(false) => {} // 没有更新或失败，继续循环
                Err(e) => log::warn!("[UPDATE] Check failed: {}", e),
            }
        }
    });
    
    Ok(())
}

/// 运行检查并安装更新
fn run_check_and_install(
    config: &Config,
    token_manager: &TokenManager,
) -> Result<bool, String> {
    let response = match check_for_update(config, token_manager) {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("[UPDATE] Check failed: {}", e);
            return Ok(false);
        }
    };
    
    if !response.has_update {
        log::info!("[UPDATE] No new version available. Current: {}", response.current_version);
        return Ok(false);
    }
    
    log::info!("[UPDATE] New version found: {} -> {}", 
        response.current_version, response.new_version);

    // 【关键】本地防护：目标版本必须高于本地实际版本才安装，
    // 避免服务端最新包版本低于本地时被降级安装旧包
    if !is_newer_version(&response.new_version, env!("CARGO_PKG_VERSION")) {
        log::info!("[UPDATE] Target version {} is not newer than local version {}, skip installation", 
            response.new_version, env!("CARGO_PKG_VERSION"));
        return Ok(false);
    }
    
    // 下载更新包
    let temp_path = match downloader::download_update(&response.download_url, &response.filename, &response.hash) {
        Ok(path) => path,
        Err(e) => {
            log::error!("[UPDATE] Download failed: {}", e);
            return Ok(false);
        }
    };
    
    log::info!("[UPDATE] Downloaded to: {}", temp_path);
    
    // 安装更新
    match installer::install_update(&temp_path) {
        Ok(_) => {
            log::info!("[UPDATE] ✅ Update completed!");
            Ok(true)
        }
        Err(e) => {
            log::error!("[UPDATE] Installation failed: {}", e);
            Ok(false)
        }
    }
}
